packages/flutter_tools/lib/src/isolated/native_assets/native_assets.dart

```diff
@@ -410,417 +565,363 @@ Uri nativeAssetsBuildUri(Uri projectUri, OS os) {
-}) async {
-  if (flutterDevices.length != 1) {
-    return dryRunNativeAssetsMultipleOSes(
-      projectUri: projectUri,
-      fileSystem: fileSystem,
-      targetPlatforms: flutterDevices.map((FlutterDevice d) => d.targetPlatform).nonNulls,
-      buildRunner: buildRunner,
-    );
+Never _throwNativeAssetsBuildDryRunFailed() {
+  throwToolExit(
+    'Building (dry run) native assets failed. See the logs for more details.',
+  );
+}
+
+Never _throwNativeAssetsBuildFailed() {
+  throwToolExit(
+    'Building native assets failed. See the logs for more details.',
+  );
+}
+
+Never _throwNativeAssetsLinkFailed() {
+  throwToolExit(
+    'Linking native assets failed. See the logs for more details.',
+  );
+}
+
+/// Select the native asset build mode for a given Flutter build mode.
+BuildModeImpl _nativeAssetsBuildMode(build_info.BuildMode buildMode) {
+  switch (buildMode) {
+    case build_info.BuildMode.debug:
+      return BuildModeImpl.debug;
+    case build_info.BuildMode.jitRelease:
+    case build_info.BuildMode.profile:
+    case build_info.BuildMode.release:
+      return BuildModeImpl.release;
   }
-  final FlutterDevice flutterDevice = flutterDevices.single;
-  final build_info.TargetPlatform targetPlatform = flutterDevice.targetPlatform!;
+}
 
-  final Uri? nativeAssetsYaml;
-  switch (targetPlatform) {
-    case build_info.TargetPlatform.darwin:
-      nativeAssetsYaml = await dryRunNativeAssetsMacOS(
-        projectUri: projectUri,
-        fileSystem: fileSystem,
-        buildRunner: buildRunner,
-      );
-    case build_info.TargetPlatform.ios:
-      nativeAssetsYaml = await dryRunNativeAssetsIOS(
-        projectUri: projectUri,
-        fileSystem: fileSystem,
-        buildRunner: buildRunner,
+Map<NativeCodeAssetImpl, KernelAsset> _assetTargetLocationsForOS(OS targetOS,
+    List<NativeCodeAssetImpl> codeAssets, bool flutterTester, Uri buildUri) {
+  switch (targetOS) {
+    case OS.windows:
+    case OS.linux:
+      final Uri? absolutePath = flutterTester ? buildUri : null;
+      return _assetTargetLocationsWindowsLinux(codeAssets, absolutePath);
+    case OS.macOS:
+      final Uri? absolutePath = flutterTester ? buildUri : null;
+      return assetTargetLocationsMacOS(codeAssets, absolutePath);
+    case OS.iOS:
+      return assetTargetLocationsIOS(codeAssets);
+    case OS.android:
+      return assetTargetLocationsAndroid(codeAssets);
+    default:
+      throw StateError('This should be unreachable.');
+  }
+}
+
+Future<void> _copyNativeCodeAssetsForOS(
+    OS targetOS,
+    Uri buildUri,
+    build_info.BuildMode buildMode,
+    FileSystem fileSystem,
+    Map<NativeCodeAssetImpl, KernelAsset> assetTargetLocations,
+    String? codesignIdentity,
+    bool flutterTester) async {
+  final List<NativeCodeAssetImpl> codeAssets =
+      assetTargetLocations.keys.toList();
+  switch (targetOS) {
+    case OS.windows:
+    case OS.linux:
+      await _copyNativeCodeAssetsToBundleOnWindowsLinux(
```

Comment:
Maybe assert that `codesignIdentity` is null. Since we don't do code signing. If we ever add code signing, we should not forget to use it here.

Reply:
Done - though I've had an assert in another place already.